modules/metrics_prometheus_v1/src/main/scala/com/evolutiongaming/skafka/metrics/KafkaMetricsCollector.scala

```diff
@@ -0,0 +1,146 @@
+  * doesn't allow such symbols in metric names and label names. Also, it ignores metrics without a name or a
+  * description. See [[https://prometheus.io/docs/practices/naming/]] for more details.
+  *
+  * Example:
+  * {{{
+  * val collectorRegistry: PrometheusRegistry = ??? // your Prometheus Java collector registry
+  * val consumer: Consumer[F, K, V]          = ??? // your consumer
+  * val collector: KafkaMetricsCollector[F]  = new KafkaMetricsCollector[F](consumer.clientMetrics)
+  * collectorRegistry.register(collector)
+  * }}}
+  */
+class KafkaMetricsCollector[F[_]: Monad: ToTry](
+  kafkaClientMetrics: F[Seq[ClientMetric[F]]],
+  prefix: Option[String] = None,
+  customLabels: List[(String, String)],
+) extends MultiCollector {
+
+  private val MetricNameRegex: Regex                 = "[a-zA-Z_:][a-zA-Z0-9_:]*".r
+  private val LabelNameRegex: Regex                  = "[a-zA-Z_][a-zA-Z0-9_]*".r
+  private val (customLabelsKeys, customLabelsValues) = customLabels.separate
+
+  override def collect(): MetricSnapshots = {
+    for {
+      metrics      <- kafkaClientMetrics
+      metricsGroups = metrics.groupBy(m => (m.name, m.group)).values.toList
+      result <- metricsGroups
+        .traverse { metricsGroup =>
+          val metric         = metricsGroup.head
+          val prometheusName = getPrometheusName(metric)
+          prometheusName match {
+            case Some(name) =>
+              metricsGroup
+                .toVector
+                .traverse(buildSample(_))
+                .map(_.flatten)
+                .map {
+                  case samples if samples.nonEmpty =>
+                    Some(buildSnapshot(name, metric.description, samples))
+
+                  case _ => Option.empty[MetricSnapshot]
+                }
+
+            case None => Option.empty[MetricSnapshot].pure
+          }
+        }
+        .map(_.flatten.asJava)
+    } yield new MetricSnapshots(result)
+  }.toTry.get
+
+  protected def getPrometheusName(metric: ClientMetric[F]): Option[String] = {
+    (metric.name, metric.description) match {
+      case (name, description) if name.nonEmpty && description.nonEmpty =>
+        val prometheusName =
+          (prefix.toList :+ metric.group :+ metric.name).mkString("_").replaceAll("-", "_")
+
+        if (MetricNameRegex.findFirstIn(prometheusName).contains(prometheusName)) prometheusName.some else None
+
+      case _ => None
+    }
+  }
+
+  private def buildSnapshot(metricName: String, description: String, samples: Vector[MetricSample]): MetricSnapshot =
+    if (metricName.endsWith("total")) buildCounterSnapshot(metricName, description, samples)
+    else buildGaugeSnapshot(metricName, description, samples)
+
+  private def buildCounterSnapshot(
+    metricName: String,
+    description: String,
+    samples: Vector[MetricSample]
+  ): MetricSnapshot = {
+    val snapshotBuilder = CounterSnapshot.builder()
```

Comment:
I'm confused on why we do this at all while not having similar code for 0.x version, could you please explain?

Reply:
The previous Prometheus client implementation allowed the creation of a Sample class that could be used with any type of metric. However, the new Prometheus client API requires snapshots to be built for each metrics type via the corresponding builder.